src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,267 @@
+    this.escapeChar = escapeChar;
+  }
+
+  /** @return Quote character specified in COPY statement, if provided. */
+  public char getQuoteChar() {
+    return this.quoteChar;
+  }
+
+  protected void setQuoteChar(char quoteChar) {
+    this.quoteChar = quoteChar;
+  }
+
+  /** @return True if copy data contains a header, false otherwise. */
+  public boolean hasHeader() {
+    return this.hasHeader;
+  }
+
+  protected void setHasHeader(boolean hasHeader) {
+    this.hasHeader = hasHeader;
+  }
+
+  public int getFormatCode() {
+    return this.formatCode;
+  }
+
+  protected void setFormatCode(int columnCount) {
+    this.formatCode = columnCount;
+  }
+
+  public MutationBuilder getMutationBuilder() {
+    return this.mutationBuilder;
+  }
+
+  protected void setMutationBuilder(MutationBuilder mutationBuilder) {
+    this.mutationBuilder = mutationBuilder;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (this.copyColumnNames.size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : this.copyColumnNames) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + this.tableName + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
+    if (columnType.matches("(i?)STRING(?:\\((?:MAX|[0-9]+)\\))?")) {
+      return TypeCode.STRING;
+    } else if (columnType.matches("(i?)BYTES(?:\\((?:MAX|[0-9]+)\\))?")) {
+      return TypeCode.BYTES;
+    } else if (columnType.equalsIgnoreCase("INT64")) {
+      return TypeCode.INT64;
+    } else if (columnType.equalsIgnoreCase("FLOAT64")) {
+      return TypeCode.FLOAT64;
+    } else if (columnType.equalsIgnoreCase("BOOL")) {
+      return TypeCode.BOOL;
+    } else if (columnType.equalsIgnoreCase("TIMESTAMP")) {
+      return TypeCode.TIMESTAMP;
+    } else {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+  }
+
+  private void queryInformationSchema() throws SQLException {
+    Map<String, TypeCode> tableColumns = new LinkedHashMap<>();
+    ResultSet result =
+        this.connection
+            .createStatement()
+            .executeQuery(
+                "SELECT "
+                    + COLUMN_NAME
+                    + ", "
+                    + SPANNER_TYPE
+                    + " FROM information_schema.columns WHERE table_name = '"
+                    + this.tableName
```

Comment:
This lookup is case sensitive. Is that intentional?

nit: It should preferably use a query parameter instead of string concatenation.

Reply:
Case sensitive table names are allowed.  I think it should be case sensitive here since this table name is read in form the user supplied query. 

I changed it to use query parameters.

Reply:
See also above: A table name in the PostgreSQL `COPY` command would normally not be case sensitive (unless it is quoted). I think it is OK to have this as a limitation in this version, but we should include a hint in the error message so the user knows what might be going on.

See https://www.postgresql.org/docs/9.4/sql-syntax-lexical.html for more information on identifier case-sensitivity in PostgreSQL. The TLDR version of the rules are:
1. Unquoted identifiers are case-insensitive. `SELECT * FROM FOO` is equivalent to `SELECT * FROM foo`.
2. Quoted identifiers are case-sensitive. `SELECT * FROM "FOO"` is selecting from a different table than `SELECT * FROM "foo"`.